### DISCUSSION TITLE
Incorrect offset and new setup file for TTGO T-Watch 2020 V3

### DISCUSSION THREAD

**Discussion by RJ-Eckie** [Bodmer/TFT_eSPI#2292] [New processor/display support or capability request]
Hi,

The output on my 240x240 ST7789 display is not properly aligned in orientations 0 and 1. Orientations 2 and 3 are correct and using the full screen. By reading your suggestions on another issue, I've worked out the correct alignments to:
```
tft.setViewport(80, 0, 240, 240); // for tft.setRotation(0);
tft.setViewport(0, 80, 240, 240); // for tft.setRotation(1);
```
However the screen is not fully filled:

```
  tft.init();
  tft.setRotation(3);
  tft.fillScreen(TFT_DARKGREY);
  tft.drawRect(0, 0, 240, 240, TFT_WHITE);
  tft.setCursor(10, 10, 4);
  tft.setTextColor(TFT_WHITE);
  tft.println("Hello World");
```
![IMG_2093](https://user-images.githubusercontent.com/121803687/210234298-aa212d42-15d3-4f68-b352-29563d469ddd.jpg)
```
  tft.init;
  tft.setRotation(0);
  tft.setViewport(80, 0, 240, 240);
  tft.fillScreen(TFT_NAVY);
  tft.drawRect(0, 0, 240, 240, TFT_CYAN);
  tft.setCursor(10, 10, 4);
  tft.setTextColor(TFT_CYAN);
  tft.println("Hello World");
```
![IMG_2094](https://user-images.githubusercontent.com/121803687/210234877-951f93c9-1a7a-41b6-ba7d-d861badbdba6.jpg)

---

There are now multiple versions of the T-Watch. (Including 2019, 2020 V1-V3 and the circular 2021)
I've adapted the general T-Watch setup for the 2020 V3:

```
#define ST7789_DRIVER

#define TFT_RGB_ORDER TFT_RGB // Red and blue are swapped without this
#define TFT_INVERSION_ON // This fixes the display showing inverted colors in some situations   

#define TFT_WIDTH  240
#define TFT_HEIGHT 240

#define TFT_BL      15
#define TFT_MISO    -1
#define TFT_MOSI    19
#define TFT_SCLK    18
#define TFT_CS      5
#define TFT_DC      27

#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts
#define SMOOTH_FONT

// #define SPI_FREQUENCY   1000000
// #define SPI_FREQUENCY   5000000
// #define SPI_FREQUENCY  10000000
// #define SPI_FREQUENCY  20000000
// #define SPI_FREQUENCY 27000000 // Actually sets it to 26.67MHz = 80/3
#define SPI_FREQUENCY  40000000
//#define SPI_FREQUENCY  80000000
```

Thanks,
RJ

**Comment by Bodmer** [comment]
The viewport is clipped to the specified screen width and height so the viewport appears clipped. To avoid this, try these settings for the setup:
#define TFT_WIDTH  240
#define TFT_HEIGHT 320

The viewport function can then be used to align the screen area. Let me know if that works and which viewport parameters correct the offset in each rotation orientation.

**Reply by RJ-Eckie** [reply]
Sadly the output does not seem to respond to TFT_WIDTH and TFT_HEIGHT at all. Setting TFT_HEIGHT (or TFT_WIDTH for that matter) to 320 doesn't remove the crop. Also, setting the values to _less_ than 240 (fe 120x120) doesn't crop the display any smaller either.

I can still flip the colors with TFT_INVERSION_ON/OFF so I know the right setup file is being used.

Any other ideas?

**Reply by Bodmer** [reply]
With:
#define TFT_WIDTH 240
#define TFT_HEIGHT 320
The offset probably needs to be in the y direction, so this should work:
```
  tft.init();
  tft.setRotation(0);
  tft.setViewport(0, 80, 240, 240);
  tft.fillScreen(TFT_DARKGREY);
  tft.drawRect(0, 0, 240, 240, TFT_WHITE);
  tft.setCursor(10, 10, 4);
  tft.setTextColor(TFT_WHITE);
  tft.println("Hello World");
```

Unfortunately the offsets are inconsistent between different board designs. If the above works then I can deduce the offset adjustments required and add a new display variant to the library.

**Reply by RJ-Eckie** [reply]
I must have messed up something in setting up everything because changing width and height didn't have any effect _at all_. Outputs were identical as before on all rotations, including the offsets.

I reinstalled the library and redid the user setup. Everything works flawlessly now with user setup 45, including defaulting to the correct inversion and RGB order. I don't see any need anymore to add a new user setup for this specific version.

Really no idea what went wrong there. Thanks for your support!
RJ

**Comment by Bodmer** [comment]
Great, glad to hear it is sorted. Thanks for letting me know.
